src/minidriver/minidriver.c

```diff
@@ -403,27 +422,37 @@ check_card_reader_status(PCARD_DATA pCardData, const char *name)
 	DWORD dwRet;
 	int r;
 
-	logprintf(pCardData, 4, "check_reader_status\n");
+	logprintf(pCardData, 1, "check_reader_status for %s\n", name);
 	if(!pCardData)
-		return SCARD_E_INVALID_PARAMETER;
+		MD_FUNC_RETURN(pCardData, 1, SCARD_E_INVALID_PARAMETER);
 
 	dwRet = check_card_status(pCardData, name);
 	if (dwRet != SCARD_S_SUCCESS)
-		return dwRet;
+		MD_FUNC_RETURN(pCardData, 3, dwRet);
 
 	vs = (VENDOR_SPECIFIC*)(pCardData->pvVendorSpecific);
 	if(!vs)
-		return SCARD_E_INVALID_PARAMETER;
+		MD_FUNC_RETURN(pCardData, 3, SCARD_E_INVALID_PARAMETER);
+
+	logprintf(pCardData, 7, "sizeof(size_t):%d sizeof(ULONG_PTR):%d sizeof(__int3264):%d sizeof pCardData->hSCardCtx:%d\n",
+		sizeof(size_t), sizeof(ULONG_PTR), sizeof(__int3264), sizeof pCardData->hSCardCtx);
 
-	logprintf(pCardData, 7, "pCardData->hSCardCtx:0x%08X hScard:0x%08X\n",
-		  (unsigned int)pCardData->hSCardCtx,
-		  (unsigned int)pCardData->hScard);
+	logprintf(pCardData, 1, "pCardData->hSCardCtx:0x%08"SC_FORMAT_LEN_SIZE_T"X hScard:0x%08"SC_FORMAT_LEN_SIZE_T"X\n",
+		(size_t)pCardData->hSCardCtx,
+		(size_t)pCardData->hScard);
 
 	if (pCardData->hSCardCtx != vs->hSCardCtx || pCardData->hScard != vs->hScard) {
-		logprintf(pCardData, 1, "HANDLES CHANGED from 0x%08X 0x%08X\n",
-			  (unsigned int)vs->hSCardCtx,
-			  (unsigned int)vs->hScard);
-		return reinit_card_for(pCardData, name);
+		logprintf(pCardData, 1, "HANDLES CHANGED from  vs->hSCardCtx:0x%08"SC_FORMAT_LEN_SIZE_T"X vs->hScard:0x%08"SC_FORMAT_LEN_SIZE_T"X\n",
+			(size_t)vs->hSCardCtx,
+			(size_t)vs->hScard);
+		if (vs->ctx) {
+			vs->hScard = pCardData->hScard;
+			vs->hSCardCtx = pCardData->hSCardCtx;
+			r = sc_ctx_use_reader(vs->ctx, &vs->hSCardCtx, &vs->hScard);
```

Comment:
I think it would be good to add a new optional `sc_reader_t *` parameter to `sc_ctx_use_reader()`. If it is passed, this reader's handles are modified; no further action is done. If this parameter is `NULL`, then `pcsc_use_reader()` can still do its dance about deleting and adding the internal readers.

Reply:
Within the minidriver there can only be one reader. `reader-pcsc.c` already has many tests for  
`priv->gpriv->cardmod`  that do not allow `SCardConnect` and  `SCardDisconnect`  (But does allow `SCardReconnect` which is probable a bug.)   
`pcsc_lock` and `pcsc_unlock`  always return `SC_SUCCESS` without doing anything because "minidriver_specs_v7.07" section 2.3.1 say:
•	A card minidriver should assume that transactions are handled by the caller, if it uses SCRM to access the card.
•	The card minidriver can assume that all entry points except CardDeleteContext are called by holding the card transaction. This cannot be assumed in CardDeleteContext because the card might have been removed or it is being called as part of a cleanup procedure.

So the whole concept of the minidriver is to use the provided handles and let the caller deal with errors.  The minidriver OpenSC has never been able to have more then one `reader_t` with different handles  because it can not do the `SCardConnect`, `SCardDisconnect`,  `SCardBeginTransaction` or `SCardEndTransaction`.

In effect  sc_ctx_use_reader() does what you ask. The test in `reader-pcsc.c` `pcsc_use_reader` 
`if (sc_ctx_get_reader_count(ctx) > 0) {` is testing if the reader_t exists. 

note that in `minidriver.c` `associate_card`   line in 7046 `sc_ctx_get_reader_count(vs->ctx)`
and line 7048 `vs->reader = sc_ctx_get_reader(vs->ctx, 0);`  list number of readers and set `vs->reader` to the first (and only) reader.

Reply:
From a functional point of view, you may be right. From a software architecture view, `sc_ctx_use_reader()` claims some sort of universal applicability otherwise we would not need to add the `use_reader` callback to every reader implementation. However, I feel that your code has too many references to the minidriver. My suggestion was intended to maintain this universality.

However, thinking about this, it may be better to simply remove the `use_reader` callback entirely and really create a function in reader_pcsc.c specifically for the interaction with the minidriver. This would simplify the overall architecture and remove some clutter from the mind of a programmer unfamiliar with the whole entity of the code.

Reply:
`pcsc_use_reader` **is** the " function in `reader_pcsc.c` specifically for the interaction with the minidriver."  

"remove some clutter from the mind of a programmer unfamiliar with the whole entity of the code." we could  use code you added:
```
fcc8ea593 (Frank Morgner        2016-11-15 22:48:48 +0100  797)         if(strcmp(ctx->app_name, "cardmod") == 0) {
fcc8ea593 (Frank Morgner        2016-11-15 22:48:48 +0100  798)                 gpriv->cardmod = 1;
fcc8ea593 (Frank Morgner        2016-11-15 22:48:48 +0100  799)         }
```
and add:
```
diff --git a/src/libopensc/reader-pcsc.c b/src/libopensc/reader-pcsc.c
index 32bf9934..7303fd5f 100644
--- a/src/libopensc/reader-pcsc.c
+++ b/src/libopensc/reader-pcsc.c
@@ -2382,6 +2382,11 @@ int pcsc_use_reader(sc_context_t *ctx, void * pcsc_context_handle, void * pcsc_c
                goto out;
        }
 
+       if (!gpriv->cardmod) {
+               ret = SC_ERROR_INTERNAL;
+               goto err;
+       }
+
        /* Only minidriver calls this and only uses one reader */
        /* if we already have a reader, use it*/
        if (sc_ctx_get_reader_count(ctx) > 0) {
```

If a programmer tried to use pcsc_use_reader, it would fail during development testing. 


0e03c6d5c 2010